File: game/forest/playing.py
import math
import random

import pygame

import global_state
from audio_helper import AudioHelper
from config import Config
from effect_type import EffectType
from forest.forest_resources import ForestResources
from forest.hurt_branch_animation import HurtBranchAnimation
from forest.hurt_flying_start import HurtFlyingStart
from forest.hurt_rock_animation import HurtRockAnimation
from forest.hurt_trap_animation import HurtTrapAnimation
from forest.win_talking import WinTalking
from game_data import GameData
from phone_events import PhoneEvents
from render_type import RenderType
from state import State


class Playing(State):
    def __init__(self, context: GameData):
        super().__init__(context)

        self.arrow_up_focus = False
        self.arrow_down_focus = False
        self.hugo_jumping_time = None
        self.hugo_crawling_time = None
        self.last_time = global_state.frame_time
        self.old_second = None
        self.HUGO_X_POS = 16

        self.needs_bottom = RenderType.PRE
        self.needs_background = RenderType.PRE

    def process_events(self, phone_events: PhoneEvents):
        if not self.arrow_up_focus and not self.arrow_down_focus:
            if phone_events.press_2:
                self.arrow_up_focus = True
                self.hugo_jumping_time = global_state.frame_time

            if phone_events.press_8:
                self.arrow_down_focus = True
                self.hugo_crawling_time = global_state.frame_time

        if self.context.forest_parallax_pos > Config.FOREST_MAX_TIME:
            self.context.forest_parallax_pos = Config.FOREST_MAX_TIME
            return WinTalking
        else:
            self.context.forest_parallax_pos += global_state.frame_time - self.last_time
            self.last_time = global_state.frame_time

        _, integer = math.modf(self.context.forest_parallax_pos)
        integer = math.floor(integer) + 1

        if integer >= Config.FOREST_MAX_TIME:
            integer = Config.FOREST_MAX_TIME - 1

        if self.old_second is None:
            self.old_second = math.floor(self.context.forest_parallax_pos)

        if self.arrow_up_focus and global_state.frame_time - self.hugo_jumping_time > 0.75:
            self.hugo_jumping_time = None
            self.arrow_up_focus = False
        if self.arrow_down_focus and global_state.frame_time - self.hugo_crawling_time > 0.75:
            self.hugo_crawling_time = None
            self.arrow_down_focus = False

        if self.old_second != math.floor(self.context.forest_parallax_pos):

            if self.context.forest_obstacles[integer] != 0 and not Config.GOD_MODE:
                if self.context.forest_obstacles[integer] == 1 and not self.arrow_up_focus:  # Catapult
                    AudioHelper.play(ForestResources.sfx_hugo_launch, self.context.audio_port)
                    AudioHelper.play(ForestResources.sfx_catapult_eject, self.context.audio_port)
                    self.context.forest_obstacles[integer] = 0
                    return HurtFlyingStart
                elif self.context.forest_obstacles[integer] == 2 and not self.arrow_up_focus:  # Trap
                    AudioHelper.play(ForestResources.sfx_hugo_hittrap, self.context.audio_port)
                    self.context.forest_obstacles[integer] = 0
                    return HurtTrapAnimation
                elif self.context.forest_obstacles[integer] == 3 and not self.arrow_up_focus:  # Rock
                    AudioHelper.play(ForestResources.sfx_hugo_hitlog, self.context.audio_port)
                    self.context.forest_obstacles[integer] = 0
                    return HurtRockAnimation
                elif self.context.forest_obstacles[integer] == 4:  # Tree
                    if self.arrow_down_focus:
                        AudioHelper.play(ForestResources.sfx_tree_swush, self.context.audio_port)
                    else:
                        AudioHelper.play(ForestResources.sfx_hugo_hitlog, self.context.audio_port)
                        self.context.forest_obstacles[integer] = 0
                        return HurtBranchAnimation

            if self.arrow_up_focus and self.context.forest_sacks[integer] != 0:
                if self.context.forest_sacks[integer] == 1:
                    self.context.forest_score += 100
                    AudioHelper.play(ForestResources.sfx_sack_normal, self.context.audio_port)
                    self.context.forest_sacks[integer] = 0
                    global_state.attacks.append((self.context.country, EffectType.SPLAT, global_state.frame_time))
                elif self.context.forest_sacks[integer] == 2:
                    self.context.forest_score += 250
                    AudioHelper.play(ForestResources.sfx_sack_bonus, self.context.audio_port)
                    self.context.forest_sacks[integer] = 0
                    global_state.attacks.append((self.context.country, EffectType.INVERT, global_state.frame_time))

        if self.get_frame_index() % 8 == 0 and not self.arrow_up_focus:
            walk_sfx = [ForestResources.sfx_hugo_walk0, ForestResources.sfx_hugo_walk1, ForestResources.sfx_hugo_walk2,
                        ForestResources.sfx_hugo_walk3, ForestResources.sfx_hugo_walk4]
            AudioHelper.play(walk_sfx[random.randint(0, 4)], self.context.audio_port)

        self.old_second = math.floor(self.context.forest_parallax_pos)

        return None

    def render(self, screen):
        fract, _ = math.modf(self.context.forest_parallax_pos)

        for index in range(len(self.context.forest_obstacles)):
            obstacle_pos = (index - self.context.forest_parallax_pos) * Config.FOREST_GROUND_SPEED

            if self.context.forest_obstacles[index] == 1:  # Catapult
                idx = self.get_frame_index() % len(ForestResources.catapult)
                dy = [45, 43, 39, 34, 29, 22, 14, 1]
                screen.blit(ForestResources.catapult[idx], (obstacle_pos - 8, 112 + dy[idx]))
            if self.context.forest_obstacles[index] == 2:  # Trap
                idx = self.get_frame_index() % len(ForestResources.trap)
                dy = [176, 173, 169, 165, 176, 176]
                screen.blit(ForestResources.trap[idx], (obstacle_pos - 8, dy[idx] - 24))
            if self.context.forest_obstacles[index] == 3:  # Rock
                idx = self.get_frame_index() % len(ForestResources.rock)
                screen.blit(ForestResources.rock[idx], (obstacle_pos - math.sin(fract * (2 * math.pi)) * 15, 120))
            if self.context.forest_obstacles[index] == 4:  # Tree
                idx = self.get_frame_index() % len(ForestResources.tree)
                screen.blit(ForestResources.lone_tree[0], (obstacle_pos - 52, -40))
                screen.blit(ForestResources.tree[idx], (obstacle_pos, 52 + 10))

        for index in range(len(self.context.forest_sacks)):
            sack_pos = (index - self.context.forest_parallax_pos) * Config.FOREST_GROUND_SPEED

            if self.context.forest_sacks[index] != 0:
                screen.blit(ForestResources.sack[0], (sack_pos-16, 32))

        for index in range(len(self.context.forest_leaves)):
            leave_pos = (index - self.context.forest_parallax_pos) * Config.FOREST_GROUND_SPEED
            if self.context.forest_leaves[index] == 1:
                screen.blit(ForestResources.leaves2[0], (leave_pos-16, -10))
            elif self.context.forest_leaves[index] == 2:
                screen.blit(ForestResources.leaves1[0], (leave_pos-16, -10))

        if self.arrow_up_focus:
            screen.blit(ForestResources.arrows[1], (256, 17))
        else:
            screen.blit(ForestResources.arrows[0], (256 + 2, 16 + 3))
        if self.arrow_down_focus:
            screen.blit(ForestResources.arrows[3], (256, 54))
        else:
            screen.blit(ForestResources.arrows[2], (256 + 2, 54 + 2))

        if self.arrow_up_focus:
            dt = (global_state.frame_time - self.hugo_jumping_time) / 0.75
            dy = -250 * dt ** 2 + 250 * dt - 22.5
            screen.blit(ForestResources.hugo_jump[self.get_frame_index() % len(ForestResources.hugo_jump)], (self.HUGO_X_POS, 40 - dy))
        elif self.arrow_down_focus:
            screen.blit(ForestResources.hugo_crawl[self.get_frame_index() % len(ForestResources.hugo_crawl)], (self.HUGO_X_POS, 105))
        else:
            screen.blit(ForestResources.hugo_side[self.get_frame_index() % len(ForestResources.hugo_side)], (self.HUGO_X_POS, 90))

File: game-web/src/forest/playing.ts
/** Port of game/forest/playing.py */

import { State, type StateClass } from '../state';
import { AudioHelper } from '../audioHelper';
import { ForestResources } from './forestResources';
import { HurtBranchAnimation } from './hurtBranchAnimation';
import { HurtFlyingStart } from './hurtFlyingStart';
import { HurtRockAnimation } from './hurtRockAnimation';
import { HurtTrapAnimation } from './hurtTrapAnimation';
import { WinTalking } from './winTalking';
import { globalState } from '../globalState';
import { Config } from '../config';
import { EffectType } from '../effectType';
import { RenderType } from '../renderType';
import type { GameData } from '../gameData';
import type { PhoneEvents } from '../phoneEvents';

export class Playing extends State {
  static readonly HINT_LOOKAHEAD = 1;
  static readonly HINT_DURATION = 1.0;

  readonly needsBackground = RenderType.PRE;
  readonly needsBottom = RenderType.PRE;

  arrowUpFocus = false;
  arrowDownFocus = false;
  hugoJumpingTime: number | null = null;
  hugoCrawlingTime: number | null = null;
  lastTime: number;
  oldSecond: number | null = null;
  readonly HUGO_X_POS = 16;

  hintShownForIndex = -1;
  hintStartTime: number | null = null;
  hintType: 'dos' | 'ocho' | null = null;

  constructor(context: GameData) {
    super(context);
    this.lastTime = globalState.frameTime;
  }

  processEvents(phoneEvents: PhoneEvents): StateClass | null {
    if (!this.arrowUpFocus && !this.arrowDownFocus) {
      if (phoneEvents.press_2) {
        this.arrowUpFocus = true;
        this.hugoJumpingTime = globalState.frameTime;
      }
      if (phoneEvents.press_8) {
        this.arrowDownFocus = true;
        this.hugoCrawlingTime = globalState.frameTime;
      }
    }

    if (this.context.forestParallaxPos > Config.FOREST_MAX_TIME) {
      this.context.forestParallaxPos = Config.FOREST_MAX_TIME;
      return WinTalking;
    } else {
      this.context.forestParallaxPos += globalState.frameTime - this.lastTime;
      this.lastTime = globalState.frameTime;
    }

    const frac = this.context.forestParallaxPos % 1;
    let integer = Math.floor(this.context.forestParallaxPos - frac) + 1;
    if (integer >= Config.FOREST_MAX_TIME) integer = Config.FOREST_MAX_TIME - 1;

    if (this.oldSecond === null) {
      this.oldSecond = Math.floor(this.context.forestParallaxPos);
    }

    // Hint system
    const hintCheckIndex = integer + Playing.HINT_LOOKAHEAD;
    if (hintCheckIndex < (this.context.forestObstacles as number[]).length && hintCheckIndex !== this.hintShownForIndex) {
      const upcomingObstacle = (this.context.forestObstacles as number[])[hintCheckIndex];
      if (upcomingObstacle !== 0) {
        const needsJump = [1, 2, 3].includes(upcomingObstacle);
        const hintType: 'dos' | 'ocho' = this.context.forestControlsInverted
          ? (needsJump ? 'ocho' : 'dos')
          : (needsJump ? 'dos' : 'ocho');

        AudioHelper.play(hintType === 'dos' ? ForestResources.sfxHintDos : ForestResources.sfxHintOcho);
        this.hintShownForIndex = hintCheckIndex;
        this.hintStartTime = globalState.frameTime;
        this.hintType = hintType;
      }
    }
    if (this.hintStartTime !== null && globalState.frameTime - this.hintStartTime > Playing.HINT_DURATION) {
      this.hintType = null;
    }

    if (this.arrowUpFocus && this.hugoJumpingTime !== null && globalState.frameTime - this.hugoJumpingTime > 0.75) {
      this.hugoJumpingTime = null;
      this.arrowUpFocus = false;
    }
    if (this.arrowDownFocus && this.hugoCrawlingTime !== null && globalState.frameTime - this.hugoCrawlingTime > 0.75) {
      this.hugoCrawlingTime = null;
      this.arrowDownFocus = false;
    }

    if (this.oldSecond !== Math.floor(this.context.forestParallaxPos)) {
      const obstacles = this.context.forestObstacles as number[];
      const sacks = this.context.forestSacks as number[];

      if (obstacles[integer] !== 0 && !Config.GOD_MODE) {
        if (obstacles[integer] === 1 && !this.arrowUpFocus) {
          AudioHelper.play(ForestResources.sfxHugoLaunch);
          AudioHelper.play(ForestResources.sfxCatapultEject);
          obstacles[integer] = 0;
          return HurtFlyingStart;
        } else if (obstacles[integer] === 2 && !this.arrowUpFocus) {
          AudioHelper.play(ForestResources.sfxHugoHittrap);
          obstacles[integer] = 0;
          return HurtTrapAnimation;
        } else if (obstacles[integer] === 3 && !this.arrowUpFocus) {
          AudioHelper.play(ForestResources.sfxHugoHitlog);
          obstacles[integer] = 0;
          return HurtRockAnimation;
        } else if (obstacles[integer] === 4) {
          if (this.arrowDownFocus) {
            AudioHelper.play(ForestResources.sfxTreeSwush);
          } else {
            AudioHelper.play(ForestResources.sfxHugoHitlog);
            obstacles[integer] = 0;
            return HurtBranchAnimation;
          }
        }
      }

      if (this.arrowUpFocus && sacks[integer] !== 0) {
        if (sacks[integer] === 1) {
          this.context.forestScore += 100;
          this.context.forestNormalSacksCollected += 1;
          AudioHelper.play(ForestResources.sfxSackNormal);
          sacks[integer] = 0;
          globalState.attacks.push([this.context.country, EffectType.SPLAT, globalState.frameTime]);
        } else if (sacks[integer] === 2) {
          this.context.forestScore += 250;
          this.context.forestGoldenSacksCollected += 1;
          AudioHelper.play(ForestResources.sfxSackBonus);
          sacks[integer] = 0;
          globalState.attacks.push([this.context.country, EffectType.INVERT, globalState.frameTime]);
        }
      }
    }

    if (this.getFrameIndex() % 8 === 0 && !this.arrowUpFocus) {
      const walks = [
        ForestResources.sfxHugoWalk0, ForestResources.sfxHugoWalk1, ForestResources.sfxHugoWalk2,
        ForestResources.sfxHugoWalk3, ForestResources.sfxHugoWalk4,
      ];
      AudioHelper.play(walks[Math.floor(Math.random() * 5)]);
    }

    this.oldSecond = Math.floor(this.context.forestParallaxPos);
    return null;
  }

  render(ctx: OffscreenCanvasRenderingContext2D): void {
    const frac = this.context.forestParallaxPos % 1;
    const obstacles = this.context.forestObstacles as number[];
    const sacks = this.context.forestSacks as number[];
    const leaves = this.context.forestLeaves as number[];

    for (let index = 0; index < obstacles.length; index++) {
      const obstaclePos = (index - this.context.forestParallaxPos) * Config.FOREST_GROUND_SPEED;

      if (obstacles[index] === 1) {
        const idx = this.getFrameIndex() % ForestResources.catapult.length;
        const dy = [45, 43, 39, 34, 29, 22, 14, 1];
        ctx.drawImage(ForestResources.catapult[idx], obstaclePos - 8, 112 + dy[idx]);
      }
      if (obstacles[index] === 2) {
        const idx = this.getFrameIndex() % ForestResources.trap.length;
        const dy = [176, 173, 169, 165, 176, 176];
        ctx.drawImage(ForestResources.trap[idx], obstaclePos - 8, dy[idx] - 24);
      }
      if (obstacles[index] === 3) {
        const idx = this.getFrameIndex() % ForestResources.rock.length;
        ctx.drawImage(ForestResources.rock[idx], obstaclePos - Math.sin(frac * 2 * Math.PI) * 15, 120);
      }
      if (obstacles[index] === 4) {
        const idx = this.getFrameIndex() % ForestResources.tree.length;
        ctx.drawImage(ForestResources.loneTree[0], obstaclePos - 52, -40);
        ctx.drawImage(ForestResources.tree[idx], obstaclePos, 52 + 10);
      }
    }

    for (let index = 0; index < sacks.length; index++) {
      const sackPos = (index - this.context.forestParallaxPos) * Config.FOREST_GROUND_SPEED;
      if (sacks[index] !== 0) {
        ctx.drawImage(ForestResources.sack[0], sackPos - 16, 32);
      }
    }

    for (let index = 0; index < leaves.length; index++) {
      const leavePos = (index - this.context.forestParallaxPos) * Config.FOREST_GROUND_SPEED;
      if (leaves[index] === 1) ctx.drawImage(ForestResources.leaves2[0], leavePos - 16, -10);
      else if (leaves[index] === 2) ctx.drawImage(ForestResources.leaves1[0], leavePos - 16, -10);
    }

    if (this.arrowUpFocus) {
      ctx.drawImage(ForestResources.arrows[1], 256, 17);
    } else {
      ctx.drawImage(ForestResources.arrows[0], 256 + 2, 16 + 3);
    }
    if (this.arrowDownFocus) {
      ctx.drawImage(ForestResources.arrows[3], 256, 54);
    } else {
      ctx.drawImage(ForestResources.arrows[2], 256 + 2, 54 + 2);
    }

    if (this.arrowUpFocus && this.hugoJumpingTime !== null) {
      const dt = (globalState.frameTime - this.hugoJumpingTime) / 0.75;
      const dy = -250 * dt ** 2 + 250 * dt - 22.5;
      ctx.drawImage(ForestResources.hugoJump[this.getFrameIndex() % ForestResources.hugoJump.length], this.HUGO_X_POS, 40 - dy);
    } else if (this.arrowDownFocus) {
      ctx.drawImage(ForestResources.hugoCrawl[this.getFrameIndex() % ForestResources.hugoCrawl.length], this.HUGO_X_POS, 105);
    } else {
      ctx.drawImage(ForestResources.hugoSide[this.getFrameIndex() % ForestResources.hugoSide.length], this.HUGO_X_POS, 90);
    }

    if (this.hintType) {
      ctx.drawImage(this.hintType === 'dos' ? ForestResources.hintOverlayDos! : ForestResources.hintOverlayOcho!, 0, 0);
    }
  }
}
